Foundation/JSONSerialization.swift

```diff
@@ -285,219 +454,176 @@ internal extension JSONSerialization {
                 case "/":
-                    writer("\\/") // U+002F solidus
+                    result += "\\/" // U+002F solidus
                 case "\u{8}":
-                    writer("\\b") // U+0008 backspace
+                    result += "\\b" // U+0008 backspace
                 case "\u{c}":
-                    writer("\\f") // U+000C form feed
+                    result += "\\f" // U+000C form feed
                 case "\n":
-                    writer("\\n") // U+000A line feed
+                    result += "\\n" // U+000A line feed
                 case "\r":
-                    writer("\\r") // U+000D carriage return
+                    result += "\\r" // U+000D carriage return
                 case "\t":
-                    writer("\\t") // U+0009 tab
+                    result += "\\t" // U+0009 tab
                 case "\u{0}"..."\u{f}":
-                    writer("\\u000\(String(scalar.value, radix: 16))") // U+0000 to U+000F
+                    result += "\\u000\(String(scalar.value, radix: 16))" // U+0000 to U+000F
                 case "\u{10}"..."\u{1f}":
-                    writer("\\u00\(String(scalar.value, radix: 16))") // U+0010 to U+001F
+                    result += "\\u00\(String(scalar.value, radix: 16))" // U+0010 to U+001F
                 default:
-                    writer(String(scalar))
+                    result += String(scalar)
             }
         }
-        writer("\"")
+        result += "\""
     }
 
-    private func serializeFloat<T: FloatingPoint & LosslessStringConvertible>(_ num: T) throws {
-        guard num.isFinite else {
-             throw NSError(domain: NSCocoaErrorDomain, code: CocoaError.propertyListReadCorrupt.rawValue, userInfo: ["NSDebugDescription" : "Invalid number value (\(num)) in JSON write"])
-        }
-        var str = num.description
+    mutating func visit<F: FloatingPoint & LosslessStringConvertible>(_ num: F) throws {
+        var str = num.description //TODO
         if str.hasSuffix(".0") {
             str.removeLast(2)
         }
-        writer(str)
+        result += str
     }
-
-    mutating func serializeNumber(_ num: NSNumber) throws {
-        if CFNumberIsFloatType(num._cfObject) {
-            try serializeFloat(num.doubleValue)
-        } else {
-            switch num._cfTypeID {
-            case CFBooleanGetTypeID():
-                writer(num.boolValue.description)
-            default:
-                writer(num.stringValue)
-            }
-        }
-    }
-
-    mutating func serializeArray(_ array: [Any?]) throws {
-        writer("[")
+    
+    mutating func visitArrayLikeSequence<Seq: Sequence>(_ array: Seq) throws {
+        result += "["
         if pretty {
-            writer("\n")
+            result += "\n"
             incAndWriteIndent()
         }
         
         var first = true
         for elem in array {
+            guard let jsonElem = elem as? _JSONValue else {
```

Comment:
Have to make sure this gets optimized out when it's called from visit(array) rather than visit(nsarray). It *should*… but it's worth double checking.